consensus-types/blocks/roblock.go

```diff
@@ -0,0 +1,85 @@
+package blocks
+
+import (
+	"sort"
+
+	fieldparams "github.com/prysmaticlabs/prysm/v4/config/fieldparams"
+	"github.com/prysmaticlabs/prysm/v4/consensus-types/interfaces"
+)
+
+// ROBlock is a value that embeds a ReadOnlySignedBeaconBlock along with its block root ([32]byte).
+// This allows the block root to be cached within a value that satisfies the ReadOnlySignedBeaconBlock interface.
+// Since the root and slot for each ROBlock is known, slices can be efficiently sorted using ROBlockSlice.
+type ROBlock struct {
+	interfaces.ReadOnlySignedBeaconBlock
+	root [32]byte
+}
+
+// Root returns the block hash_tree_root for the embedded ReadOnlySignedBeaconBlock.Block().
+func (b ROBlock) Root() [32]byte {
+	return b.root
+}
+
+// NewROBlockWithRoot creates an ROBlock embedding the given block with its root. It accepts the root as parameter rather than
+// computing it internally, because in some cases a block is retrieved by its root and recomputing it is a waste.
+func NewROBlockWithRoot(b interfaces.ReadOnlySignedBeaconBlock, root [32]byte) (ROBlock, error) {
+	if err := BeaconBlockIsNil(b); err != nil {
+		return ROBlock{}, err
+	}
+	return ROBlock{ReadOnlySignedBeaconBlock: b, root: root}, nil
+}
+
+// NewROBlock creates a ROBlock from a ReadOnlySignedBeaconBlock. It uses the HashTreeRoot method of the given
+// ReadOnlySignedBeaconBlock.Block to compute the cached root.
+func NewROBlock(b interfaces.ReadOnlySignedBeaconBlock) (ROBlock, error) {
+	if err := BeaconBlockIsNil(b); err != nil {
+		return ROBlock{}, err
+	}
+	root, err := b.Block().HashTreeRoot()
+	if err != nil {
+		return ROBlock{}, err
+	}
+	return ROBlock{ReadOnlySignedBeaconBlock: b, root: root}, nil
+}
+
+// ROBlockSlice implements sort.Interface so that slices of ROBlocks can be easily sorted.
+// A slice of ROBlock is sorted first by slot, with ties broken by cached block roots.
+type ROBlockSlice []ROBlock
+
+var _ sort.Interface = ROBlockSlice{}
```

Comment:
This is great. I'd love to see some tests on the sort implementation

Reply:
:+1: added `roblock_test.go`